### PR TITLE
Debug updated parameter in Closure in statemachine

### PR DESCRIPTION
**Pull request by antonssonj**
I stumbled upon this bug, at first I didn't realise that it was only the debugger that gave me the wrong result. It was pretty confusing so I wanted to go to the bottom with it.
 
Steps to reproduce
  1. Create a new method that either is async or contains yield return
  2. The method should have a parameter that is used by a lambda expression in the body
  3. Modify the parameter and set a breakpoint
  4. In the debugger the parameters value is not updated
 
Found an issue for it https://github.com/dotnet/roslyn/issues/10649
 
The fix updates the displayclass variable if the variable is a parameter and the displayclass kind is a LambdaDisplayClass.
 
There is a problem with a UnitTest. Is it really supposed to be like that? When I compile the code in the UnitTest and look at it in ildasm the only way that x is used is through the closure. Never through the statemachine. And when I look at the code that does the rewrite I see that the lambda rewrite happens first and all references to the parameter in the body is rewritten to the closure and later when the statemachine is created the references are rewritten to be through the statemachines closure field.
 
So I have not added any new UnitTest but changed the test that I think is wrong.

### CODE REVIEW

**Comment by antonssonj** [comment]
Don't think the failed test is caused by my change. How do I run the tests again?

**Comment by antonssonj** [comment]
@pilche can you take a look?

**Comment by tmat** [comment]
@KevinH-MS @cston

**Comment by KevinH-MS** [comment]
Awesome--thanks much, @antonssonj!  Change LGTM.

The change to the unit test looks correct to me (the previous baseline was incorrect).

**Comment by KevinH-MS** [comment]
@cston any thoughts 👍 or 👎 ?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs [hunk 1]
```diff
@@ -1522,16 +1522,17 @@ static void G(Action a)
...
                 System.Runtime.CompilerServices.TaskAwaiter V_2,
                 C.<M>d__2 V_3,
                 System.Exception V_4)
   IL_0000:  ldarg.0
-  IL_0001:  ldfld      ""int C.<M>d__2.x""
-  IL_0006:  ret
+  IL_0001:  ldfld      ""C.<>c__DisplayClass2_0 C.<M>d__2.<>8__1""
+  IL_0006:  ldfld      ""int C.<>c__DisplayClass2_0.x""
+  IL_000b:  ret
 }");
```
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/LocalsTests.cs]: Please add ```[WorkItem(10649, "https://github.com/dotnet/roslyn/issues/10649")]``` to this test method.

**Comment by cston** [comment]
Looks like the same issue exists in VB with:
```
Imports System
Imports System.Threading.Tasks
Class C
    Shared Sub F(s As String, a As Action(Of String))
        a(s)
    End Sub
    Shared Async Function M(s As String) As Task
        Await Task.Yield()
        s = "2"
        F("2",
          Sub(t)
              If t = s Then
              End If
          End Sub)
    End Function
    Shared Sub Main()
        M("1").Wait()
    End Sub
End Class
```
And perhaps with a similar fix at https://github.com/cston/roslyn/blob/master/src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/CompilationContext.vb#L1286.

**Comment by cston** [comment]
Looks like the same issue exists in VB with:
```
Imports System
Imports System.Threading.Tasks
Class C
    Shared Sub F(s As String, a As Action(Of String))
        a(s)
    End Sub
    Shared Async Function M(s As String) As Task
        Await Task.Yield()
        s = "2"
        F("2",
          Sub(t)
              If t = s Then
              End If
          End Sub)
    End Function
    Shared Sub Main()
        M("1").Wait()
    End Sub
End Class
```
And perhaps with a similar fix at https://github.com/dotnet/roslyn/blob/master/src/ExpressionEvaluator/VisualBasic/Source/ExpressionCompiler/CompilationContext.vb#L1286.

**Comment by cston** [comment]
@antonssonj Thanks for fixing this issue. We can log a separate issue for VB, it does not need to be addressed here.
This change LGTM.

**Comment by KevinH-MS** [comment]
I've filed #12219 for the VB issue.  I'll add the work item attribute in an additional commit.  @antonssonj thanks again!

**Pull request merged**